aws/client.go

```diff
@@ -84,32 +105,46 @@ func (t *ResilientTransport) tries(req *http.Request) (res *http.Response, err e
 		buf.ReadFrom(req.Body)
 	}
 
-	for try := 0; try < t.MaxTries; try += 1 {
+	// deadline for retrying
+	retryingDeadline := time.Now().Add(t.RetryingTimeout)
+	retries := 0
+	for { // Watch the infinite loop here
 		// Each retry should use a copy of the body.
 		// This fixes a bug where subsequent retries would be using a reader
 		// that was already read.
 		if resetBody {
 			req.Body = ioutil.NopCloser(bytes.NewReader(buf.Bytes()))
 		}
+		logRequest(req)
 		res, err = t.transport.RoundTrip(req)
 
 		if !t.ShouldRetry(req, res, err) {
+			logResponse(res)
 			break
 		}
+
 		if res != nil {
 			res.Body.Close()
 		}
-		if t.Wait != nil {
-			t.Wait(try)
+
+		if retryingDeadline.Sub(time.Now()) > 0 && t.Wait != nil {
+			t.Wait(retries, t.MinRetryWait, t.MaxRetryWait)
+			retries += 1
+			continue
 		}
+		break
 	}
-
 	return
 }
 
-func ExpBackoff(try int) {
-	time.Sleep(100 * time.Millisecond *
-		time.Duration(math.Exp2(float64(try))))
+func ExpBackoff(try int, minWait time.Duration, maxWait time.Duration) {
+	wait := time.Duration(math.Pow(2, float64(try))) * minWait
+	if wait < minWait || wait > maxWait {
```

Comment:
IIRC, it can be less that minWait in case of an overflow while computing the exponential backoff.